### PR TITLE
[ENH] Added piecewise_poisson

### PR DESCRIPTION
**Pull request by Pyyim**
Co-authored-by: Kevin <kevinlamdata@gmail.com>

#### Reference Issues/PRs

Reference # 3363

#### What does this implement/fix? Explain your changes.

Added to datagen.py in preparation for adding to PoissonHMM

#### Does your contribution introduce a new dependency? If yes, which one?

No

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.
- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

### CODE REVIEW

File: sktime/annotation/datagen.py
Comment (miraep8) [inline sktime/annotation/datagen.py:271]: Example needs to be updated for poisson!

File: sktime/annotation/datagen.py
Comment (miraep8) [inline sktime/annotation/datagen.py:258]: Small change - but I would recommend you call this "lamba" rather than "frequency" (just to follow the typical syntax seen with Poisson distributions) and explain that it is the expected number (and variance! Aren't Poisson distributions cool?) of events within each time interval/data point.

**Review by miraep8 — COMMENTED** [review]
I have a few small change requests - but overall looks good! May I also request that if either of you are not in our contributors list already you go ahead and add yourselves? :)

**Comment by miraep8** [comment]
Adding a quick test for this under sktime -> annotation -> tests would also not be such a bad idea!

File: sktime/annotation/datagen.py
Comment (lmmentel) [reply sktime/annotation/datagen.py:258]: This is one point where python and maths don't play well since `lambda` is a reserved keyword in python. I've been in this situation before and still don't have a good solution for an alternative name to something everyone knows should be called `lambda`.

File: sktime/annotation/datagen.py
Comment (lmmentel) [inline sktime/annotation/datagen.py:281]: Would you mind turning `assert` into a `try`/`except` block raising human readable errors? The reason is that in some cases of production code there is a way to skip assert statements entirely (see e.g. this [link](https://realpython.com/python-assert-statement/#understanding-the-__debug__-built-in-constant)). Thanks!

**Review by lmmentel — CHANGES_REQUESTED** [review]

**Comment by lmmentel** [comment]
Thanks @Pyyim :raised_hands: All looks good except for the small comment I made on the `assert` statement that could be turned into a try/except to be more robust.

File: sktime/annotation/datagen.py
Comment (miraep8) [reply sktime/annotation/datagen.py:258]: Oh whoops, of course!  Perhaps in this case `lambdas` would work (since it is an array).  And then could use `lam` or similar when iterating over it.

**Comment by Pyyim** [comment]
Thank you for the quick response. We will get to everything you have mentioned and resolve them by tomorrow.

**Comment by Pyyim** [comment]
@miraep8 @lmmentel we have submitted the changes to the files that were requested. 

- included tests in a new file test_datagen.py
- updated examples, included the try/except block, and changed name of frequency to lambdas in datagen.py
- added ourselves to the .all-contributorsrc file

**Comment by miraep8** [comment]
Excellent!  Thank you for submitting tests for `data_gen` as well! Unfortunately there seem to be some slight issues with the tests (can check out the logs in the CI/CD to see what exactly is failing.  I think it looks like an easy-to resolve issue though! :) 

Thanks again!

**Comment by Pyyim** [comment]
@miraep8 Just finished fixing the issue you just mentioned. Thank you!

**Comment by miraep8** [comment]
Hey @Pyyim, I think there are still some merge conflicts with this branch! :)  I think once those are resolved we will be in a much better place to merge this.

**Comment by Pyyim** [comment]
@miraep8 Sorry about that. I believe it should be good to go now.

**Review by miraep8 — DISMISSED** [review]

**Comment by miraep8** [comment]
Nice work!  🎉 Unfortunately running into the issue with the small merge conflict from the contributors file again.  I tried to change it on my end but was running into some permissions issues, so perhaps might be better for you to fix!  I am planning to also merge the PR from your colleague @JonathanBechtel today as well (once tests pass) so it could just be worth waiting until that event, then resolving the merge conflicts.  Sorry - an unfortunate issue with lots of new contributors!

**Comment by miraep8** [comment]
The other PR from a new contributor has been merged!  I hope now you should be able to get the merge conflict fixed with no more errors and we can hopefully quickly get this merged.  Thank you for your patience and your work on this! :)

**Comment by Pyyim** [comment]
Hi, I currently don't have my computer right now as today is a holiday in the US, but I should be able to get to it first thing in the morning tomorrow. Thank you!

**Comment by miraep8** [comment]
No worries! The "quickly" was more on myself - ie once you resolve the issue I will try to merge it quickly to avoid any future merge conflicts!  :)  Enjoy the day off!

**Comment by Pyyim** [comment]
Hi @miraep8, 
Would it be possible to remove the all contributor file for this PR and create a separate PR for just for that?

**Comment by miraep8** [comment]
If you do that we could then re-run the test suite on just your code changes and once those pass (sure they will based on the above) we can merge this and add your contribution separately! :)

**Comment by Pyyim** [comment]
@miraep8 Would it actually be okay to create a new pull request? I believe I might have messed up with removing the all-contributors file, and it might be cleaner.

**Comment by miraep8** [comment]
No worries :)  That works!

**Pull request closed without merging**
